protocol/l1-upgrades.md

```diff
@@ -0,0 +1,338 @@
+upgraded (requiring only new values to be set).
+
+# Problem Statement + Context
+
+Our upgrades have the following issues:
+
+1. They are divorced from the implementation, meaning the work of writing the upgrade script occurs
+   in `superchain-ops` apart from the work of writing the updated code in the `optimism` monorepo.
+2. The act of writing the upgrade script is onerous, requiring:
+   1. A bespoke script with custom assertions.
+   2. A bespoke Validations.md document
+3. The work required to perform the upgrade is difficult for signers, as they must work through the
+   Validations.md document. This work has taken up to an hour in the past.
+4. The effort described above will scale linearly with the number of OP Chains.
+
+# Proposed Solution
+
+Each new release will have its own OP Contracts Manager which can perform the following distinct
+actions:
+
+1. Deploy new proxies for a new OP Chain (this work is already covered by [link to Blaine's doc]())
+1. Upgrade all existing OP Chains.
+1. Deploy any newly added superchain-shared contracts.
+1. Upgrade existing superchain-shared contracts.
+
+Note that the above functionality also makes for a good set of milestones, each of which can and
+should be implemented incrementally.
+
+## Declarative storage management
+
+Taking inspiration from Chugsplash, we wish to separate the concern of what bytecode to run, and
+what to write to storage. However we also wish to avoid using the current `StorageSetter` pattern,
+which simply provides logic to "write anything anywhere" in storage.
+
+In order to achieve this, each contract will have its `initializer()` function moved intoa
+release specific `__Populator` contract.
+
+By way of example, we take an initializable version of `SimpleStorage`:
+
+- the current version of the contract has only a `uint256 foo` variable.
+- the new version of the contract introduces a `bytes32 bar` variable.
+
+The status quo implementation of the new Simple Storage would look like this:
+
+```solidity
+contract SimpleStorage is Initializable {
+  uint public foo;
+  bytes32 public bar; // new variable added for the current release
+
+  constructor() {
+    _disableInitializers();
+  }
+
+  /// @notice initialize is called whether for a fresh system or when upgrading an existing system.
+  function initialize(uint _foo, bytes32 _bar) public initializer {
+    // chains being upgraded need to read the current value of foo, then write it again.
+    foo = _foo;
+    bar = _bar;
+  }
+}
+```
+
+This document proposes that the above contract should be refactored accordingly:
+
+```solidity
+/// @notice The base contract with all runtime logic
+contract SimpleStorage {
+  uint public foo;
+  bytes32 public bar; // new variable added for the current release.
+}
+
+/// @notice an ephemeral contract which is temporarily set as the implementation during deployment
+///         or upgrae.
+contract SimpleStoragePopulator is SimpleStorage {
```

Comment:
This isn't guaranteed to work due to codesize limits. This will not work with the `OptimismPortal` today as any new code will push it over the limit. So it would mean that we would have to pay off tech debt before being able to do this sometimes

Reply:
This is a very lazy approach in my opinion where we opt towards higher deploy time/compile time costs under the guise of simplicity. This introduces tech debt that we will need to clean up in the future, we should remove an intermediate step and just go towards a more simple approach first.

Thinking of these options as better approaches:
- use storage setter, we already have tests for the constructor/initializer working as expected
- refactor the storage layout to be its own contract so that the populator doesn't need the implementation details

Reply:
I'd be wary of using a storage setter - feels very error prone. Storage layout being in its own contract & inheriting from that is probably a bit better

Reply:
We are no longer using Populator, so I won't bother to defend it here. 
Touching on other topics: 

- Using a base storage layout contract: I don't think this is as simple or low overhead as it sounds. For example we would have the following inheritance graph: `L1CrossDomainMessenger -> CrossDomainMessenger  -> CrossDomainMessengerLayout`. This would mean that the `L2CrossDomainMessenger` can no longer inherit from `CrossDomainMessenger `, which is something we should strive to maintain.
- Using `StorageSetter`: I think we can use this pattern quite safely by: 
  1. Placing the `initialized` value in the same unstructured storage location for all contracts. 
  2. Using a custom contract (in place of `StorageSetter`) to reset that slot prior to each upgrade.